### ISSUE TITLE
Extensions: pre-compilation fails if more than one extension

### ISSUE THREAD

**Issue by omlins** [JuliaLang/julia#52511]
MWE: https://github.com/omlins/ParallelStencil.jl/pull/130/files#diff-4fd906dc9bd8d9fc59a9eab475a5119bbd4505ba46a091bede7d8cec6cabe8b8
In brief this MWE contains:
```toml
name = "Diffusion3D_minimal"
uuid = "91637811-9d05-4227-81a2-fe1d069b6ecb"
authors = ["Samuel Omlin <samuel.omlin@cscs.ch>"]
version = "0.1.0"

[deps]
ParallelStencil = "94395366-693c-11ea-3b26-d9b7aac5d958"

[compat]
ParallelStencil = ">= 0.4.0"

[weakdeps]
AMDGPU = "21141c5a-9bdb-4563-92ae-f87d6854732e"
CUDA = "052768ef-5323-5732-b1bb-66c8b64840ba"

[extensions]
D3D_AMDGPUExt = "AMDGPU"
D3D_CUDAExt = "CUDA"
```
and:
```julia
module Diffusion3D_minimal
    using ParallelStencil
    f() = return
end

module D3D_CUDAExt
    import Diffusion3D_minimal
    g() = return
end

module D3D_AMDGPUExt
    import Diffusion3D_minimal
    h() = return
end
```
The error is the following:
```
julia> Base.compilecache(Base.PkgId(Base.UUID("94395366-693c-11ea-3b26-d9b7aac5d958"), "ParallelStencil")); Base.compilecache(Base.PkgId(Base.UUID("91637811-9d05-4227-81a2-fe1d069b6ecb"), "Diffusion3D_minimal")); Base.compilecache(Base.PkgId(Base.UUID("38fae68b-b861-5d3c-b423-1a3f5200d4e2"), "D3D_CUDAExt"))
[ Info: Precompiling ParallelStencil [94395366-693c-11ea-3b26-d9b7aac5d958]
[ Info: Precompiling Diffusion3D_minimal [91637811-9d05-4227-81a2-fe1d069b6ecb]
[ Info: Precompiling D3D_CUDAExt [38fae68b-b861-5d3c-b423-1a3f5200d4e2]
┌ Warning: Module D3D_CUDAExt with build ID ffffffff-ffff-ffff-000e-c267534cc71f is missing from the cache.
│ This may mean D3D_CUDAExt [38fae68b-b861-5d3c-b423-1a3f5200d4e2] does not support precompilation but is imported by a module that does.
└ @ Base loading.jl:1942
┌ Error: Error during loading of extension D3D_CUDAExt of Diffusion3D_minimal, use `Base.retry_load_extensions()` to retry.
│   exception =
│    1-element ExceptionStack:
│    Declaring __precompile__(false) is not allowed in files that are being precompiled.
│    Stacktrace:
│      [1] _require(pkg::Base.PkgId, env::Nothing)
│        @ Base ./loading.jl:1946
│      [2] __require_prelocked(uuidkey::Base.PkgId, env::Nothing)
│        @ Base ./loading.jl:1806
│      [3] #invoke_in_world#3
│        @ Base ./essentials.jl:921 [inlined]
│      [4] invoke_in_world
│        @ Base ./essentials.jl:918 [inlined]
│      [5] _require_prelocked
│        @ Base ./loading.jl:1797 [inlined]
│      [6] _require_prelocked
│        @ Base ./loading.jl:1796 [inlined]
│      [7] run_extension_callbacks(extid::Base.ExtensionId)
│        @ Base ./loading.jl:1289
│      [8] run_extension_callbacks(pkgid::Base.PkgId)
│        @ Base ./loading.jl:1324
│      [9] run_package_callbacks(modkey::Base.PkgId)
│        @ Base ./loading.jl:1158
│     [10] __require_prelocked(uuidkey::Base.PkgId, env::String)
│        @ Base ./loading.jl:1813
│     [11] #invoke_in_world#3
│        @ Base ./essentials.jl:921 [inlined]
│     [12] invoke_in_world
│        @ Base ./essentials.jl:918 [inlined]
│     [13] _require_prelocked(uuidkey::Base.PkgId, env::String)
│        @ Base ./loading.jl:1797
│     [14] macro expansion
│        @ Base ./loading.jl:1784 [inlined]
│     [15] macro expansion
│        @ Base ./lock.jl:267 [inlined]
│     [16] __require(into::Module, mod::Symbol)
│        @ Base ./loading.jl:1747
│     [17] #invoke_in_world#3
│        @ Base ./essentials.jl:921 [inlined]
│     [18] invoke_in_world
│        @ Base ./essentials.jl:918 [inlined]
│     [19] require(into::Module, mod::Symbol)
│        @ Base ./loading.jl:1740
│     [20] include
│        @ Base ./Base.jl:489 [inlined]
│     [21] include_package_for_output(pkg::Base.PkgId, input::String, depot_path::Vector{String}, dl_load_path::Vector{String}, load_path::Vector{String}, concrete_deps::Vector{Pair{Base.PkgId, UInt128}}, source::String)
│        @ Base ./loading.jl:2216
│     [22] top-level scope
│        @ stdin:3
│     [23] eval
│        @ Core ./boot.jl:383 [inlined]
│     [24] include_string(mapexpr::typeof(identity), mod::Module, code::String, filename::String)
│        @ Base ./loading.jl:2070
│     [25] include_string
│        @ Base ./loading.jl:2080 [inlined]
│     [26] exec_options(opts::Base.JLOptions)
│        @ Base ./client.jl:316
│     [27] _start()
│        @ Base ./client.jl:552
└ @ Base loading.jl:1295
("/scratch/snx3000tds/julia/omlins/daint-gpu/compiled/v1.10/D3D_CUDAExt/owmDu_K34O8.ji", "/scratch/snx3000tds/julia/omlins/daint-gpu/compiled/v1.10/D3D_CUDAExt/owmDu_K34O8.so")
```

If one of the two extensions is deactivated by commenting the corresponding line in the `project.toml` (e.g. `#D3D_AMDGPUExt = "AMDGPU"` to deactivate the `AMDGPU` extension), then the problem disappears, e.g.:
```
julia> Base.compilecache(Base.PkgId(Base.UUID("94395366-693c-11ea-3b26-d9b7aac5d958"), "ParallelStencil")); Base.compilecache(Base.PkgId(Base.UUID("91637811-9d05-4227-81a2-fe1d069b6ecb"), "Diffusion3D_minimal")); Base.compilecache(Base.PkgId(Base.UUID("38fae68b-b861-5d3c-b423-1a3f5200d4e2"), "D3D_CUDAExt"))
[ Info: Precompiling ParallelStencil [94395366-693c-11ea-3b26-d9b7aac5d958]
[ Info: Precompiling Diffusion3D_minimal [91637811-9d05-4227-81a2-fe1d069b6ecb]
[ Info: Precompiling D3D_CUDAExt [38fae68b-b861-5d3c-b423-1a3f5200d4e2]
("/scratch/snx3000tds/julia/omlins/daint-gpu/compiled/v1.10/D3D_CUDAExt/owmDu_K34O8.ji", "/scratch/snx3000tds/julia/omlins/daint-gpu/compiled/v1.10/D3D_CUDAExt/owmDu_K34O8.so")
```

Note that ParallelStencil does not contain any module `__init__` definition.

The problem appears with both tested julia versions, 1.9.4 and 1.10.

**Comment by KristofferC** [comment]
There is not really any Pkg code being executed here so this repo is actually probably better.

**Comment by omlins** [comment]
@KristofferC  Do you have any idea what is wrong here?

**Comment by omlins** [comment]
@KristofferC : could the reason for the failure be that ParallelStencil itself does not yet use extensions and has both GPU packages as normal dependencies?

**Comment by KristofferC** [comment]
I have not looked into this one yet.

**Comment by lassepe** [comment]
Duplicate of https://github.com/JuliaLang/julia/issues/52132?

**Comment by KristofferC** [comment]
I don't think so. That's related to stdlibs in the sysimage.

**Comment by ViralBShah** [comment]
I ran into this when just doing `Pkg.test("Sundials")`.

**Comment by alecloudenback** [comment]
Update: I deleted my `packages` and `compiled` subfolders in my `.julia` and I don't see the issue anymore.

**Comment by KristofferC** [comment]
I looked at this now and this is identical to https://github.com/JuliaLang/julia/issues/53081. The issue is the following:

- ParallelStencil depends on AMDGPU and CUDA.
- The package depends on ParallelStencil and has extensions that gets loaded from AMDGPU and CUDA:

So what happens when the package loads:

- Package starts loading
- It loads ParallelStencil
- ParallelStencil loads AMDGPU
- Our extension for AMDGPU starts loading
  - The extension loads the package
  - The package loads ParallelStencil
  - ParallelStencil loads CUDA
  - The extension for CUDA starts loading
    - The extension loads the package
    - The package loads ParallelStencil
    - ParallelStencil loads AMDGPU
    -  The extension for AMDGPU starts loading

As you see, we are now in a cycle where the extension ends up loading each other back and forth. This will happen whenever:

- You have two or more extensions
- The triggers for these extensions are in the set of your transitive dependencies.

https://github.com/JuliaLang/julia/pull/53112 was an attempt to fix this but I have heard some uncertainty from other people if that way is correct.

**Comment by omlins** [comment]
@KristofferC : thanks a lot for looking into this.

ParallelStencil does now not anymore have AMDGPU and CUDA as fixed dependencies, but as weak dependencies - the corresponding functionality has been moved into extensions (merged into main: https://github.com/omlins/ParallelStencil.jl/pull/144).  Nevertheless, the issue remains the same. 
@KristofferC : Does your explanation from above also apply to this case?

**Comment by KristofferC** [comment]
On https://github.com/JuliaLang/julia/pull/53972 this prints

```
┌ Warning: Circular dependency detected. Precompilation will be skipped for:
│   Base.PkgId(Base.UUID("47e5d35c-c2d0-5408-acd2-db5e4eb7937b"), "MolSimStylePlotting")
│   Base.PkgId(Base.UUID("e2da7e33-71d8-5845-bdbe-2073996eaa79"), "REMDPlotting")
│   Base.PkgId(Base.UUID("169be7d3-c844-588e-8a3f-7557bbfa2dc9"), "BlockAveragesPlotting")
```

**Comment by lmiq** [comment]
The extensions work without problems after loading. 

I don't see any circular dependency there. In fact, the extensions depend on `MolSimToolkit` itself (the parent package) and on `Plots`.

**Comment by KristofferC** [comment]
All three of the extensions seem to have Plots as their trigger. So the cycle is the following:


1. BlockAveragesPlotting loads MolSimToolkit + Plots
2. This causes MolSimStylePlotting to load.
3. MolSimStylePlotting  loads MolSimToolkit + Plots
4. This causes BlockAveragesPlotting to load
5. Go to 1.

I think you should fold all these into one extension.

**Comment by lmiq** [comment]
That's not clear from my development perspective, as "loading the extension" is not something one does manually in any moment, as what we do is to load the trigger packages. Probably some clarification is needed there.

**Comment by KristofferC** [comment]
An extension gets loaded when all its triggers (+ parent) get loaded. So if you have two extensions that both load the parent and the triggers they will try to load each other. 

https://github.com/JuliaLang/julia/pull/53112 tries to stop this but it is unclear what is the best and right thing to do in situations like this.

**Comment by jakobjpeters** [comment]
I believe that I also have this issue, as seen in this [CI log](https://github.com/jakobjpeters/Typstry.jl/actions/runs/9624785062/job/26548948857).

**Comment by topolarity** [comment]
There are some important fixes coming in 1.10.8:
  - https://github.com/JuliaLang/julia/pull/56666 fixes a lot of the error messages above
  - https://github.com/JuliaLang/julia/pull/55910 fixes many false cycles

It will be good to get a read on this issue once 1.10.8 is out

**Comment by topolarity** [comment]
> This issue is still present in 1.10.8:

Yeah, this specific case was brought up in https://github.com/JuliaObjects/Accessors.jl/issues/192#issuecomment-2581423150

All _false_ cycles (that we know of) are fixed on 1.10, but unfortunately it still suffers from a few _true_ cycles.

The problem is that `Accessors → LinearAlgebraExt` and `Accessors → DatesExt` both effectively have a trigger set of `[Accessors, BaseSysimage]` since Dates/LinearAlgebra are in the sysimage.

That means that pre-compiling `Accessors → LinearAlgebraExt` will trigger `Accessors → DatesExt` to load and vice-versa, unless we make a breaking change on 1.10 to fix that by design (see https://github.com/JuliaLang/julia/pull/55589/ for the 1.11 change, with fallout/justification in https://github.com/JuliaLang/julia/issues/56204#issuecomment-2419175793). Accessors.jl should have worked around this in https://github.com/JuliaObjects/Accessors.jl/pull/193, but it is still a problem for custom sysimages, etc. that can end up putting many triggers into the sysimage, leading to a "cycle" again.

An alternative fix here would be to make extensions load less eagerly against things in the sysimage (https://github.com/JuliaLang/julia/commit/08d229f4a7cb0c3ef5becddd1b3bc4f8f178b8e4). In that case, the two extensions above don't load each other just because they are in the sysimage - you need to `using LinearAlgebra` etc. to trigger the extension. @KristofferC tried this out earlier, but it ended up being reverted due to buggy loading behavior (see https://github.com/JuliaLang/julia/pull/56234)

**Comment by aplavin** [comment]
Just checking: there's still no way to have multiple stdlib extensions if one wants the package to compile on Julia 1.10?
Encountered again in https://github.com/JuliaMath/IntervalSets.jl/issues/206.

Makes me wonder how long 1.10 will remain lts then... :)

**Comment by wsmoses** [comment]
@oscardssmith per our conversation earlier today, was the fix here backported to 1.10?

**Comment by topolarity** [comment]
Yeah, this was backported to 1.10 in https://github.com/JuliaLang/julia/pull/60997

You might be worried about https://github.com/EnzymeAD/Reactant.jl/issues/2764 though - that fix (https://github.com/JuliaLang/julia/pull/61591) has been sitting in the release queue for a while since the last Julia 1.10 release was in March.
